osquery/tables/system/linux/process_open_pipes.cpp

```diff
@@ -0,0 +1,193 @@
+    return nullptr;
+  return createPipeInfoStruct(std::stoi(process),
+                              std::stoi(desc.first),
+                              getMode(process, desc.first),
+                              file_stat.st_ino,
+                              "named");
+}
+
+std::unique_ptr<pipe_info> getPipeInfo(
+    const std::string& process,
+    const std::pair<std::string, std::string>& desc) {
+  if (desc.second.find("pipe:") != std::string::npos) // found unnamed pipe
+    return createPipeInfoStruct(std::stoi(process),
+                                std::stoi(desc.first),
+                                getMode(process, desc.first),
+                                parseInode(desc.second),
+                                "anonymous");
+  else // check for potential named pipe
+    return getNamedPipeInfo(process, desc);
+}
+
+void genPipePartners(const std::string& process,
+                     const std::map<std::string, std::string>& descriptors,
+                     PidToPipesMap& pipe_desc,
+                     InodeToPipesMap& pipe_partners,
+                     std::vector<std::unique_ptr<pipe_info>>& pipe_structs) {
+  std::unique_ptr<pipe_info> ps;
+  for (const auto& desc : descriptors) {
+    ps = getPipeInfo(process, desc);
+    if (!ps)
+      continue;
+    pipe_desc[ps->pid].push_back(ps.get());
+    pipe_partners[ps->inode].push_back(ps.get());
+    pipe_structs.push_back(std::move(ps));
+  }
+}
+
+void createRow(const std::string& process,
+               const pipe_info* ps,
+               const pipe_info* partner_ps,
+               QueryData& results) {
+  Row r;
+
+  r["pid"] = process;
+  r["fd"] = std::to_string(ps->fd);
+  r["mode"] = ps->mode;
+  r["inode"] = std::to_string(ps->inode);
+  r["type"] = ps->type;
+
+  if (partner_ps) {
+    r["partner_pid"] = std::to_string(partner_ps->pid);
+    r["partner_fd"] = std::to_string(partner_ps->fd);
+    r["partner_mode"] = partner_ps->mode;
+  }
+
+  results.push_back(r);
+}
+
+void genResults(const std::string& process,
+                const PidToPipesMap& pipe_desc,
+                const InodeToPipesMap& pipe_partners,
+                QueryData& results) {
+  auto pipe_desc_iter = pipe_desc.find((std::stoi(process)));
+  if (pipe_desc_iter == pipe_desc.end())
+    return;
+
+  for (const auto& ps :
+       pipe_desc_iter->second) { // iterate over vector of pipe_info
+    for (const auto& partner_ps : pipe_partners.at(ps->inode)) {
+      if (isUnconnectedPipe(ps->inode, pipe_partners))
+        createRow(process, ps, nullptr, results);
+      else if (isSamePipe(ps, partner_ps) ||
+               !isReaderWriterPair(ps, partner_ps))
+        continue;
+      else
+        createRow(process, ps, partner_ps, results);
+    }
+  }
+}
+
+QueryData genPipes(QueryContext& context) {
+  QueryData results;
+  std::set<std::string> pids;
+  PidToPipesMap pipe_desc;
+  InodeToPipesMap pipe_partners;
+  std::vector<std::unique_ptr<pipe_info>> pipe_structs;
```

Comment:
How is `pipe_structs` used, can you narrow the scope?

Reply:
`pipe_structs` is used to keep alive the scope of the `unique_ptr`s, the references to which are used in `pipe_desc` and `pipe_partners`.